Changelog 154 (#132)
intellij-rust/intellij-rust.github.io: _posts/2021-08-30-changelog-154.markdown

date: 2021-08-30 13:00:00 +0300
---

## New Features

* [#7640] Improve external ABI support:
    - Parse any literal in ABI name position (including raw strings) and annotate non-string literals as invalid (in the same way as `rustc` does)

      PIC
    - Annotate [invalid ABI](https://doc.rust-lang.org/error-index.html#E0703) names and experimental ABIs without the corresponding feature attribute

      PIC
    - Provide competition for ABI names

      GIF

* [#7682] Detect [E0121](https://doc.rust-lang.org/error-index.html#E0121) in structs and enum variants (by [@t-kameyama])

    GIF

* [#7646] Implement `Expand into separate table` intention to expand an TOML inline table into full one

    GIF

* [#7666] Show [macro 2.0](https://rust-lang.github.io/rfcs/1584-macros.html) items in [Structure View](https://www.jetbrains.com/help/idea/viewing-structure-of-a-source-file.html) (`Ctrl` + `F12`)

* [#7664] Improve parser recovery for [function pointers](https://doc.rust-lang.org/reference/types/function-pointer.html) (by [@afetisov])

## Performance Improvements

* [#7703] Suggest keywords much more quickly during Rust code completion

## Fixes

> [comment]: <> (* [#7711] Introduce `Ty.isEquivalentTo&#40;Ty&#41;` for `Ty`s comparison. Unlike `== `, it ignores `aliasedBy`)

Reviewer: Let's remove it. There are user-visible fixes, but the explanation will take too much of my time. I don't think it's worth it.

[comment]: <> ( field.)
* [#7635], [#7676] Take into account type aliases in most of [intention actions, quick fixes](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-code-generation.html#intention-actions), and [refactorings](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-refactorings.html) (by [@Kobzol])

* [#7557] Reduce amount of false positive annotations caused by `Self convention` inspection (by [@Kobzol])

> * [#7692] Don't wrap associated types by type aliases in `Chained method` inlay hints (by [@Kobzol])

Reviewer: In theory, this affects all inlay type hints, not only `Chained method` hints. But it's true that this is most useful for `Chained method` hints

* [#7690] Import types after invoking `Create function`, `Create struct` and `Create tuple struct` intentions (by [@Kobzol])

* [#7638] `Surround with unsafe` inspection now works identically inside statements and tail expressions

* [#7598] Display [implied shortcut reference links](https://rust-lang.github.io/rfcs/1946-intra-rustdoc-links.html#implied-shortcut-reference-links) in quick documentation (by [@shocoman])

* [#7582] Don't show a completion popup automatically when typing a `let binding` identifier starting from a lowercase letter.

> You still can invoke the completion explicitly (using `Ctrl`+`Enter`, for example)

Reviewer 2: Nice catch!

  <img src="/assets/posts/changelog-154/type-placeholder.png" alt="The type placeholder `_` is not allowed within types on item signatures for structs and enum variants" width="700px">

* [#7646] Implement `Expand into separate table` intention to expand an inline TOML table into a full one

  {% include gif-img.html path="/assets/posts/changelog-154/expand-into-separate-table" w="700px" %}

* [#7666] Show [macro 2.0](https://rust-lang.github.io/rfcs/1584-macros.html) items in [Structure View](https://www.jetbrains.com/help/idea/viewing-structure-of-a-source-file.html) (`Ctrl`+`F12`)

* [#7664] Improve parser recovery for [function pointers](https://doc.rust-lang.org/reference/types/function-pointer.html) (by [@afetisov])

## Performance Improvements

* [#7703] Suggest keywords much more quickly during Rust code completion

## Fixes

* [#7635], [#7676] Take into account type aliases in most of [intention actions, quick fixes](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-code-generation.html#intention-actions), and [refactorings](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-refactorings.html) (by [@Kobzol])

> * [#7557] Reduce amount of false positive annotations caused by `Self convention` inspection (by [@Kobzol])

Reviewer 3: ```suggestion
* [#7557] Reduce the number of false-positive annotations caused by `Self convention` inspection (by [@Kobzol])
```

> * [#7692] Don't wrap associated types by type aliases in inlay hints including `Chained method` ones (by [@Kobzol])

Reviewer 3: ```suggestion
* [#7692] Do not substitute type aliases for associated types in inlay hints like `Chained method` (by [@Kobzol])
```
Not sure about it

* [#7690] Import types after invoking `Create function`, `Create struct` and `Create tuple struct` intentions (by [@Kobzol])

* [#7711] Fix some quick-fixes and intentions to make them work with aliased types properly.

>   For example, `` Convert to String using `ToString` trait `` quick-fix works now even if expected type is alias for `String` type

Reviewer 3: ```suggestion
  For example, `` Convert to String using `ToString` trait `` quick-fix now works even when the expected type is an alias for `String` type
```
